python/ray/tune/impl/placeholder.py

```diff
@@ -0,0 +1,248 @@
+            for i, choice in enumerate(config.categories)
+        ]
+        return config
+    elif isinstance(config, Function):
+        # Function type.
+        id_hash = _id_hash(id_prefix)
+        v = _FunctionResolver(id_hash, config)
+        resolvers[path_prefix].append(v)
+        return v.get_placeholder()
+    elif not isinstance(config, Domain):
+        # Other non-search space reference objects, dataset, actor handle, etc.
+        id_hash = _id_hash(id_prefix)
+        v = _RefResolver(id_hash, config)
+        resolvers[path_prefix].append(v)
+        return v.get_placeholder()
+    else:
+        # All the other cases, do nothing.
+        return config
+
+
+def _get_placeholder(config: Any, prefix: Tuple, path: Tuple):
+    if not path:
+        return prefix, config
+
+    key = path[0]
+    if (isinstance(config, dict) and key in config) or (
+        isinstance(config, list) and key < len(config)
+    ):
+        # Expand config tree recursively.
+        return _get_placeholder(config[key], prefix=prefix + (path[0],), path=path[1:])
+
+    if isinstance(config, tuple):
+        if config[0] in (_FunctionResolver.TOKEN, _RefResolver.TOKEN):
+            # Found a matching placeholder.
+            # Note that we do not require that the full path are consumed before
+            # declaring a match. Because this placeholder may be part of a nested
+            # search space. For example, the following config:
+            #   config = {
+            #       "param1": tune.grid_search([
+            #           tune.grid_search([Object1, 2, 3]),
+            #           tune.grid_search([Object2, 5, 6]),
+            #       ]),
+            #   }
+            # will result in placeholders under path ("param1", 0, 0).
+            # After expansion though, the choosen placeholder will live under path
+            # ("param1", 0) like this: config = {"param1": (Placeholder1, 2, 3)}
+            return prefix, config
+        elif key < len(config):
+            return _get_placeholder(
+                config[key], prefix=prefix + (path[0],), path=path[1:]
+            )
+
+    # Can not find a matching placeholder.
+    return None, None
+
+
+@DeveloperAPI
+def resolve_placeholders(config: Any, replaced: defaultdict):
+    """Replaces placeholders contained by a config dict with the original values.
+
+    Args:
+        config: The config to replace placeholders in.
+        replaced: A dict from path to replaced objects.
+    """
+
+    def __resolve(resolver_type, args):
+        for path, resolvers in replaced.items():
+            if not resolvers:
+                # Should never happen.
+                continue
+
+            if not isinstance(resolvers[0], resolver_type):
+                continue
+
+            prefix, ph = _get_placeholder(config, (), path)
+            if not ph:
+                # Represents an unchosen value. Just skip.
+                continue
+
+            for resolver in resolvers:
```

Comment:
Can we just make `resolvers` a hash map, where key is `resolver.hash`? Currently we have linear search with respect to the search space size, which can be huge.

Reply:
I think the list of options shouldn't be long man. this is probably fine, and looks simpler.
thanks.